### ISSUE TITLE
Handling timeouts when server is unreachable

### ISSUE THREAD

**Issue by mrcoles** [niwinz/django-redis#155]
I’m using:

*   django-redis==4.2.0
*   redis==2.10.3
*   Django==1.7.7
*   Python 3.4.2

I’m testing out what happens when I give django-redis a bad URL endpoint for the redis cache (emulating what would happen with a complete failure) and it appears to take a lot longer than the "SOCKET_CONNECT_TIMEOUT" and "SOCKET_TIMEOUT" settings would suggest. For example I have a connect timeout of 2 and socket timeout of 0.5, but it takes 16 seconds before failing to connect, which happens lazily during every single lookup.

I tracked down some particulars and extracted the following info at various points in the chain.

```python
# in django_redis.pool.ConnectionFactory.make_connection_params
kwargs = {
    'url': 'redis://bad-cache-url:6379/0',
    'socket_timeout': 0.5,
    'socket_connect_timeout': 2,
    'parser_class': <class 'redis.connection.PythonParser'>,
}

# client is django_redis.client.default.DefaultClient

# which uses redis.StrictRedis
```

I was testing this out using `python3 manage.py shell` with the following simple call:

```python
from django.core.cache import cache
cache.get('test')
```

I noticed:

1. `redis.StrictRedis.execute_command` has a baked in retry

2. compounded by the above call, `redis.connection.Connection._connect` tries 4 times (looks like `socket.getaddrinfo(self.host, self.port, 0, socket.SOCK_STREAM)` returns 4 items).

Given the connect timeout of 2 * execute_command tries of 2 * connect tries of 4 = 16 seconds—my observed 16 seconds now appears to make sense.

So… two questions:

1.  Should the docs be more clear that a SOCKET_CONNECT_TIMEOUT is per connection attempt and that there could be 8 connection attempts (maybe more in other configurations?)

2.  Is there a smarter way django-redis could handle failed connections? It seems really wasteful to keep trying to connect when the host is unreachable / down / bad dns / etc. In high performance environments, there’s usually a concept of a “circuit breaker” (Netflix uses this terminology, some other folks call it “markdown”), where after enough fails on connects or calls, the client will immediately fail on any future calls, which degrades the experience, but prevents web servers from reaching thread exhaustion because of too many long-running requests. Also, it could use exponential backoff or some form of small sampling to periodically retry and self-repair if things ever came back up. This would be a really nice feature. I suppose it’d mostly be useful for particular use cases where the cache is more additive and not mission-critical to an application and when the user is using `DJANGO_REDIS_IGNORE_EXCEPTIONS = True`.

Thoughts?

**Comment by niwinz** [comment]
Hmm, very strange. The redis-py default behavior is not retry on timeout exception is raised: https://github.com/andymccurdy/redis-py/blob/4d0b0afe9c9a431ed50c3e9fb95a0aa88b1f1038/redis/client.py#L568 so your behavior of several retries it is not very normal. Maybe if you point a redis client to not existent server it raises a connectionerror instead of timeouterror?

We can expose more redis-py settings for make it more flexible but at this point we can't do anything if a problem is a redis-py. Can you revise the version of redis-py that you are using?

About the other question, in the past it there an other plugable client that handles some kind of failover. That in case of timeout switches for some time to a fallback redis server. It was removed for no time to maintain it. But this kind of behavior can be easyly implemented on top of default client using the django-redis pluggable client approach.

**Comment by mrcoles** [comment]
Good point on the pluggable client approach!

What do you mean by revising the version of redis-py that I’m using? The one that auto-installed when I installed `django-redis` was `redis==2.10.3`. LMK if I should try something different.

In practice, the retries aren’t a bad idea, since sporadic shit happens, but in this scenario I was curious about the massive failure case and avoiding a long connect loop every single time. I’ll let you know if I put together a pluggable client—it would have to probably use local memory and be per-client, which is not ideal at all, but could work (unless there’s another way to access shared memory across a server without I/O calls). I suppose the other option is treating the cache as mission critical…

**Comment by niwinz** [comment]
I asked the question about the redis-py version just for check if a installed version is very different but thanks for confirm that you are using the latest one.

As I said previously, the retries happens because the exception raised is not the `TimeoutException`, I hope is `ConnectionError`. Because you are pointing to not existent host. If you point to the existent point but with redis server is very busy the `TimeoutException` will be raised. I thinking is so...

Let me know I can help you or you consider that django-redis needs some improvements.

**Comment by robnardo** [comment]
Hi niwinz & mrcoles, 

Thank you both for adding your feedback to this bug / issue.  I am also having this issue. I am using python 2.7, django 1.7.9, django-redis 4.2.0 and redis 2.10.3

I have **DJANGO_REDIS_IGNORE_EXCEPTIONS = True** and 

```
'OPTIONS': {
         'COMPRESS_MIN_LEN': 10,
         'SOCKET_CONNECT_TIMEOUT': 2,
         'SOCKET_TIMEOUT': 2,
         'CLIENT_CLASS': 'django_redis.client.DefaultClient',
         'TIMEOUT': 5,
     },
```

but I still get the thrown exception.

The error I get is **ConnectionError at ... Error 11001 connecting to redis.domain.com:6380. getaddrinfo failed.** And the error is located at: **django_redis\cache.py in _decorator, line 30**

I hope that helps. Just wondering, do have a planned fix for this coming up?

**Comment by mrcoles** [comment]
@robnardo looks like you’re facing a different issue? If you’re still dealing with it and see it as a potential bug in django_redis, then you should make a separate issue.

**Comment by niwinz** [comment]
@mrcoles :+1:

**Comment by kylemacfarlane** [comment]
It seems like there was a bug that was fixed in redis-py 2.10.6. Connection timeouts are now raised as `TimeoutError`.

However the issue this ticket was created for still stands. The problem is that every time the cache is accessed django_redis calls methods on the underlying `StrictRedis` which fires off a reconnection attempt. So if your `SOCKET_CONNECT_TIMEOUT` is set to 1 and you call `cache.get()` 20 times during a request then django_redis will still spend 20x 1s trying to connect to Redis.

I think a more sensible approach would be to only try to connect to Redis once per request or once every 30-60s.

**Comment by Havko** [comment]
This is exactly what I was looking for.  This works very well.
